components/popover/utils.js

```diff
@@ -0,0 +1,86 @@
+
+/**
+ * Module constants
+ */
+const HEIGHT_OFFSET = 10; // used by the arrow and a bit of empty space
+const isMobileViewport = () => window.innerWidth < 782;
+
+/**
+ * Utilitity used to compute the popover position and the content max width/height for a popover
+ * given its anchor rect and its content size.
+ *
+ * @param {Object} anchorRect       Anchor Rect.
+ * @param {Object} contentSize      Content Size.
+ * @param {string} position         Position.
+ * @param {boolean} expandOnMobile  Whether to expand the popover on mobile or not.
+ *
+ * @return {Object} Popover position and constraints.
+ */
+export function computePopoverPosition( anchorRect, contentSize, position = 'top', expandOnMobile = false ) {
+	const { width, height } = contentSize;
+	const popoverLeft = Math.round( anchorRect.left + ( anchorRect.width / 2 ) );
+	const [ yAxis, xAxis = 'center' ] = position.split( ' ' );
+
+	// y axis aligment choices
+	const topAlignment = {
+		popoverTop: anchorRect.top,
+		contentHeight: anchorRect.top - HEIGHT_OFFSET - height > 0 ? height : anchorRect.top - HEIGHT_OFFSET,
+	};
+	const bottomAlignment = {
+		popoverTop: anchorRect.bottom,
+		contentHeight: anchorRect.bottom + HEIGHT_OFFSET + height > window.innerHeight ? window.innerHeight - HEIGHT_OFFSET - anchorRect.bottom : height,
+	};
+
+	// x axis alignment choices
+	const centerAlignment = {
+		contentWidth: (
+			( popoverLeft - ( width / 2 ) > 0 ? ( width / 2 ) : popoverLeft ) +
+			( popoverLeft + ( width / 2 ) > window.innerWidth ? window.innerWidth - popoverLeft : ( width / 2 ) )
+		),
+	};
+	const leftAlignment = {
+		contentWidth: popoverLeft - width > 0 ? width : popoverLeft,
+	};
+	const rightAlignment = {
+		contentWidth: popoverLeft + width > window.innerWidth ? window.innerWidth - popoverLeft : width,
+	};
+
+	// Choosing the y axis
+	let chosenYAxis;
+	let contentHeight = null;
+	if ( yAxis === 'top' && topAlignment.contentHeight === height ) {
+		chosenYAxis = 'top';
+	} else if ( yAxis === 'bottom' && bottomAlignment.contentHeight === height ) {
+		chosenYAxis = 'bottom';
+	} else {
+		chosenYAxis = topAlignment.contentHeight > bottomAlignment.contentHeight ? 'top' : 'bottom';
+		const chosenHeight = chosenYAxis === 'top' ? topAlignment.contentHeight : bottomAlignment.contentHeight;
+		contentHeight = chosenHeight !== height ? chosenHeight : null;
+	}
+
+	// Choosing the x axis
+	let chosenXAxis;
+	let contentWidth = null;
+	if ( xAxis === 'center' && centerAlignment.contentWidth === width ) {
```

Comment:
Minor: Maybe we can use a switch.

On the switches we would do:
```
case 'center':
	chosenXAxis = centerAlignment.contentWidth === width ? 'center' : undefined;
```

After the switch, we would check if chosenXAxis was undefined if yes we would use our last else. It is just a personal preference normally I prefer switches instead of a series of else.

It may also make sense to extract some parts of the function into smaller ones e.g: one part that deals with the x and other that deals with y, just to make computePopoverPosition smaller :)

Reply:
I don't like the switch in this case personally, I like "dumb" code :P but I like the splitting proposal which I did.